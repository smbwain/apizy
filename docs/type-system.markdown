---
layout: page
title: Type system
permalink: /type-system/
nav_order: 3
---

# type-system
{: .no_toc }

apizy type-system is used to describe types of your API data structures. Input and output.

- TOC
{:toc}

# In short, how does it work

An _apizy type-system_ is based on special _helpers_: javaScript functions that define types.

It allows _apizy_ to extract appropriate types from your definitions on compilation stage, as well as do runtime checks of entire input and output when running server and also generate typescript definitions for SDK.

E.g.

```ts
import {createApi, object, float, arrayOf, optional, string, nullable, uuid, boolean} from 'apizy';
const api = createApi();

api.createEntity(
    'methodName',
    object({
        a: float(),
        b: optional(string()),
        c: nullable(arrayOf(object({
            id: uuid(),
        }))),
    }),
    boolean(),
    async (input) => {
        // input has type {
        //      a: number;
        //      b?: string;
        //      c: Array<{
        //          id: string;
        //      }> | null;
        // }
    }
);
```

# Helpers

## Primitive types

### string

```ts
string()
```

### int

```ts
int()
```

### float

```ts
float()
```

### boolean

```ts
boolean()
```

### uuid

```ts
uuid()
```

### oneOf

For fields that must match one of several values, use the `oneOf` helper.

```ts
const userRole = oneOf(['admin', 'editor', 'subscriber']);

const userShape = object({
    id: uuid(),
    role: userRole,
});
```

This ensures the `role` field only accepts one of the predefined values.

## Structures

### object

Defines an object that uses multiple types for its fields,

```ts
object({
    id: uuid(),
    email: string(),
    privateData: object({
        firstName: string(),
        lastName: string(),
    }),
});
```

### arrayOf

```ts
const tagsShape = arrayOf(string());

const blogPostShape = object({
    title: string(),
    tags: tagsShape,
});
```

This ensures `tags` is an array of strings.

## Modifiers

### nullable

When a field can be `null`, use the `nullable` helper.

```ts
const userShape = object({
    id: uuid(),
    phoneNumber: nullable(string()),
});
```

`phoneNumber` can either be a string or `null`.

### optional

To allow a field to be optional (can be omitted in the input), use the `optional` helper.

```ts
const userShape = object({
    id: uuid(),
    nickname: optional(string()),
});
```

Here `nickname` may or may not be included in the object.

### defaultValue

The same as [optional](#optional), but allows to set default value.

```ts
const userShape = object({
    role: defaultValue('user', string()),
});
```

Here if the `role` field is not provided, it defaults to `'user'`.

## Relations

### extend

This helper allows you to add parts of output, which should not be loaded by default.
But it can be loaded (extended) only if a client app requested that.

You can do that, because of performance reason, or size of the object.

**Example**

```ts
import {createApi, object, float, arrayOf, optional, string, nullable, uuid, boolean} from 'apizy';
const api = createApi();

api.createEntity(
    'users.getById',
    string(),
    object({
        id: uuid(),
        name: string(),
        topArticles: extend(arrayOf(object({  // <-- extend helpers here
            id: uuid(),
            text: string(),
        }))),
    }),
    async (input) => {
        return {
            id: '...',
            name: '...',
            topArticles: async () => {
                // this function will be called to load articles,
                // only when user request this extension
                return [ {
                    id: '...',
                    text: '...',
                } /* ,...*/ ];
            },
        };
    }
);
```

On client side:
```ts
sdk.users.getById('...')
// Returns: Promise<{
//     id: string;
//     name: string;
// }>
```

```ts
sdk.users.getById('...', {
    extend: {
        topArticles: {},
    },
});
// Returns: Promise<{
//     id: string;
//     name: string;
//     topArticles: {
//         id: string;
//         text: string;
//     },
// }>
```

Extends can be nested.

### relation

TODO

> Check out [Entities](./api#entities) first.