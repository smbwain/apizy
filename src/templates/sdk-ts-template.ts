import {
    ApiDescription,
    MethodDescription, TypeDescription,
} from '../index';

export const sdkTsTemplate = (desc: ApiDescription) => `

export type Method<Input, Output, ExtendQuery, FetchOptions> = (
    input: Input,
    options: {
        extend: ExtendQuery;
        fetchOptions: FetchOptions;
    },
) => Promise<Output>;

${Object.keys(desc.types).map(name => {
    const d = desc.types[name];
    return `export type T_${name} = ${renderTSTypeFromDescription(d, 'output')}\nexport interface ExtendQuery_${name} ${renderTSExtendQueryTypeFromDescription(d)}`;
}).join('\n\n')}

export interface SDK<FetchOptions> {
${renderMethodsTree(desc.methods, 'type', 1)}}

export function createSDK<FetchOptions>(
    fetch: (methodName: string, input: any, extend?: any, fetchOptions?: FetchOptions) => Promise<any>,
): SDK<FetchOptions> {
    const createMethod = (methodName: string) => (input: any, options: any) => fetch(methodName, input, options?.extend, options?.fetchOptions);
    return {\n${renderMethodsTree(desc.methods, 'impl', 2)}    };
};
`;

const tab = (indent: number) => ' '.repeat(indent*4);

const renderTSTypeFromDescription = (
    desc: TypeDescription,
    mode: 'input' | 'output',
    indent: number = 0,
): string => {
    if (desc.type.ts) {
        return desc.type.ts;
    }
    if (desc.type.alias) {
        return `T_${desc.type.alias}`;
    }
    if (desc.type.arrayOf) {
        return `Array<${renderTSTypeFromDescription(desc.type.arrayOf, mode, indent)}>`;
    }
    if (desc.type.objectOf) {
        return `{\n${Object.keys(desc.type.objectOf).map(key => {
            const sub = desc.type.objectOf![key];
            return `${tab(indent+1)}${JSON.stringify(key)}`+
                `${((mode === 'input' && sub.inputOptional) || (mode === 'output' && sub.outputExtendable)) ? '?' : ''}`+
                `: ${renderTSTypeFromDescription(sub, mode, indent+1)};\n`;
        }).join('')}${tab(indent)}}`;
    }
    throw new Error('Unknown type');
};

const renderTSExtendQueryTypeFromDescription = (
    desc: TypeDescription,
    indent: number = 0,
): string => {
    if (desc.type.arrayOf) {
        return renderTSExtendQueryTypeFromDescription(desc.type.arrayOf, indent);
    }
    if (desc.type.alias) {
        return `ExtendQuery_${desc.type.alias}`;
    }
    if (desc.type.objectOf) {
        const keys = Object.keys(desc.type.objectOf).filter(key => desc.type.objectOf![key].outputExtendable);
        if (keys.length > 0) {
            return `{\n${keys.map(key => {
                const sub = desc.type.objectOf![key];
                return `${tab(indent+1)}${JSON.stringify(key)}?: ${renderTSExtendQueryTypeFromDescription(sub, indent+1)};\n`;
            }).join('')}${tab(indent)}}`;
        }
    }
    return '{}';
};

// const renderMethodType = (desc: MethodDescription, indent: number): string => {
//     return `(\n`+
//         `${tab(indent+1)}input: ${desc.input ? renderTSTypeFromDescription(desc.input, 'input', indent+1) : 'null'},\n` +
//         `${tab(indent+1)}options: {\n${tab(indent+2)}extend: ${renderTSExtendQueryTypeFromDescription(desc.output, indent+2)};\n${tab(indent+2)}fetchOptions: FetchOptions;\n${tab(indent+1)}},\n` +
//         `${tab(indent)}) => Promise<${renderTSTypeFromDescription(desc.output, 'output', indent+1)}>`;
// };

const renderMethodType = (desc: MethodDescription, indent: number): string => {
    return `Method<${desc.input ? renderTSTypeFromDescription(desc.input, 'input', indent) : 'null'}, ` +
        `${renderTSTypeFromDescription(desc.output, 'output', indent)}, ` +
        `${renderTSExtendQueryTypeFromDescription(desc.output, indent)}, FetchOptions>`;
};

const renderMethodsTree = (rec: Record<string, MethodDescription>, mode: 'type' | 'impl', indent: number): string => {
    let lastPath: string[] = [];
    const parts: string[] = [];
    const close = (c: number, indent: number) => {
        for (; c > 0; c--) {
            parts.push(`${tab(indent+c-1)}}${mode === 'type' ? ';' : ','}\n`);
        }
    };
    Object.keys(rec).sort().forEach(key => {
        const path = key.split('.');
        const methodName = path.pop();

        let commonCount = 0;
        while (commonCount < lastPath.length && commonCount < path.length && lastPath[commonCount] === path[commonCount]) {
            commonCount++;
        }

        close(lastPath.length-commonCount, indent+path.length-1);
        lastPath = path;

        for(let i = commonCount; i < path.length; i++) {
            parts.push(`${tab(indent+path.length-1)}${JSON.stringify(path[i])}: {\n`);
        }

        parts.push(`${tab(indent+path.length)}${JSON.stringify(methodName)}: `);
        if (mode === 'type') {
            parts.push(`${renderMethodType(rec[key], indent+path.length)};\n`);
        } else {
            parts.push(`createMethod(${JSON.stringify(key)}),\n`);
        }
    });
    close(lastPath.length, indent);
    return parts.join('');
};