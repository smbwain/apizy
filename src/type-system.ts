import * as check from 'checkeasy';
import {mapObject} from './utils';
import {TypeDescription} from './api-description';

export type PromiseOrValue<T> = Promise<T> | T;

export interface InputType<T> {
    extract: check.Validator<T>;
    desc: TypeDescription;
}

export interface OutputType<T> {
    resolve: (v: T, path: string, extendQuery: any, ctx: any) => PromiseOrValue<any>;
    checkExtendedQuery?: check.Validator<any>;
    desc: TypeDescription;
}

export interface Type<T> extends InputType<T>, OutputType<T> {}

export function isTypeInput<T>(t: InputType<T> | OutputType<T>): t is InputType<T> {
    return !!(t as any).extract;
}
export function isTypeOutput<T>(t: InputType<T> | OutputType<T>): t is OutputType<T> {
    return !!(t as any).resolve;
}

export function extend<T>(type: OutputType<T>): OutputType<() => PromiseOrValue<T>> {
    return {
        desc: {
            ...type.desc,
            outputExtendable: true,
        },
        resolve: async (v, path, extendQuery, ctx): Promise<any> => {
            if (!extendQuery) {
                return undefined;
            }
            return type.resolve(
                await v(),
                path,
                extendQuery,
                ctx,
            );
        },
        checkExtendedQuery: type.checkExtendedQuery || check.optional(check.object({})),
    };
}

export const string = (options?: Parameters<typeof check.string>[0]): Type<string> => ({
    extract: check.string(options),
    resolve: check.string(options),
    desc: {
        type: {
            ts: 'string',
        },
    },
});

export const uuid = (): Type<string> => ({
    extract: check.UUID(),
    resolve: check.UUID(),
    desc: {
        type: {
            ts: 'string',
        },
    },
});

export const int = (options?: Parameters<typeof check.int>[0]): Type<number> => ({
    extract: check.int(options),
    resolve: check.int(options),
    desc: {
        type: {
            ts: 'number',
        },
    },
});

export const float = (options?: Parameters<typeof check.float>[0]): Type<number> => ({
    extract: check.float(options),
    resolve: check.float(options),
    desc: {
        type: {
            ts: 'number',
        },
    },
});

export const boolean = (): Type<boolean> => ({
    extract: check.boolean(),
    resolve: check.boolean(),
    desc: {
        type: {
            ts: 'boolean',
        },
    },
});

// export function json<T>(scheme: check.Validator<T>): Type<T> {
//     return {
//         extract: (v, path) => {
//             let parsed;
//             try {
//                 parsed = JSON.parse(v);
//             } catch(err) {
//                 throw new check.ValidationError(`Not valid JSON not valid for [${path}]`);
//             }
//             return scheme(parsed, path);
//         },
//         resolve: (v) => JSON.stringify(v),
//         tsExtract: 'string',
//         tsResolve: 'string',
//     };
// }

export function object<O extends {[k: string]: Type<any>}>(description: O): Type<{
    [K in keyof O]: O[K] extends Type<infer T> ? T : never;
}>;
export function object<O extends {[k: string]: InputType<any>}>(description: O): InputType<{
    [K in keyof O]: O[K] extends InputType<infer T> ? T : never;
}>;
export function object<O extends {[k: string]: OutputType<any>}>(description: O): OutputType<{
    [K in keyof O]: O[K] extends OutputType<infer T> ? T : never;
}>;
export function object(description: any) {
    const keys = Object.keys(description);
    const res: any = {
        desc: {
            type: {
                objectOf: mapObject(description, (val: any) => val.desc),
            },
        },
    };
    if (keys.every(key => isTypeInput(description[key]))) {
        const d = description as Record<string, InputType<any>>;
        (res as InputType<any>).extract = check.object(mapObject(d, (val: any) => val.extract));
    }
    if (keys.every(key => isTypeOutput(description[key]))) {
        const d = description as Record<string, OutputType<any>>;
        (res as OutputType<any>).resolve = async (v, path, extendQuery, ctx) => {
            const res: any = {};
            await Promise.all(keys.map(async (key) => {
                res[key] = await d[key].resolve(v[key], `${path}.${key}`, extendQuery?.[key], ctx);
            }));
            return res;
        };

        const subValidators: Record<string, check.Validator<any>> = {};
        for (const key of keys) {
            if (d[key].checkExtendedQuery) {
                subValidators[key] = d[key].checkExtendedQuery!;
            }
        }

        (res as OutputType<any>).checkExtendedQuery = Object.keys(subValidators).length ? check.optional(check.object(subValidators)) : undefined;
    }
    return res;
}

export function arrayOf<T>(t: Type<T>, options?: Parameters<typeof check.arrayOf>[1]): Type<T[]>;
export function arrayOf<T>(t: InputType<T>, options?: Parameters<typeof check.arrayOf>[1]): InputType<T[]>;
export function arrayOf<T>(t: OutputType<T>, options?: Parameters<typeof check.arrayOf>[1]): OutputType<T[]>;
export function arrayOf(t: Type<any> | InputType<any> | OutputType<any>, options?: Parameters<typeof check.arrayOf>[1]) {
    const res: any = {
        desc: {
            type: {
                arrayOf: t.desc,
            },
        },
    };
    if (isTypeInput(t)) {
        (res as InputType<any>).extract = check.arrayOf(t.extract, options);
    }
    if (isTypeOutput(t)) {
        (res as OutputType<any>).resolve = async (v, path, extendedQuery, ctx) => {
            if (t.desc.outputExtendable && !extendedQuery) {
                return undefined;
            }
            return Promise.all((v as any[]).map((item, index) => t.resolve(item, `${path}[${index}]`, extendedQuery, ctx)))
        };
        (res as OutputType<any>).checkExtendedQuery = t.checkExtendedQuery;
    }
    return res;
}

export function nullable<T>(t: Type<T>): Type<T | null>;
export function nullable<T>(t: InputType<T>): InputType<T | null>;
export function nullable<T>(t: OutputType<T>): OutputType<T | null>;
export function nullable(t: Type<any> | InputType<any> | OutputType<any>) {
    const res: any = {
        desc: {
            ...t.desc,
            nullable: true,
        },
    };
    if (isTypeInput(t)) {
        (res as InputType<any>).extract = check.nullable(t.extract);
    }
    if (isTypeOutput(t)) {
        (res as OutputType<any>).resolve = async (v, path, extendedQuery, ctx) => {
            if (t.desc.outputExtendable && !extendedQuery) {
                return undefined;
            }
            if (v === null) {
                return null;
            }
            return t.resolve(v, path, extendedQuery, ctx);
        };
        (res as OutputType<any>).checkExtendedQuery = t.checkExtendedQuery;
    }
    return res;
}

export const optional = <T>(t: InputType<T>): InputType<T | undefined> => ({
    desc: {
        ...t.desc,
        inputOptional: true,
    },
    extract: check.optional(t.extract),
});

export const defaultValue = <T>(val: T, t: InputType<T>): InputType<T> => ({
    desc: {
        ...t.desc,
        inputOptional: true,
    },
    extract: check.defaultValue(val, t.extract),
});

export const oneOf = <T>(values: ReadonlyArray<T>): Type<T> => ({
    desc: {
        type: {
            ts: `${values.map(val => JSON.stringify(val)).join(' | ')}`,
        }
    },
    extract: check.oneOf(values),
    resolve: check.oneOf(values),
});

export const any = (): Type<any> => ({
    desc: {
        type: {
            ts: 'any',
        },
    },
    extract: v => v,
    resolve: v => v,
});

export const exact = <T>(value: T): Type<T> => ({
    desc: {
        type: {
            ts: JSON.stringify(value),
        },
    },
    extract: check.exact(value),
    resolve: check.exact(value),
});

export function relation<T>(
    type: OutputType<T>,
    load: (id: string, ctx: any) => Promise<T | null>,
    options: {
        nullable: true;
        auto?: boolean;
    },
): OutputType<string | null>;
export function relation<T>(
    type: OutputType<T>,
    load: (id: string, ctx: any) => Promise<T>,
    options?: {
        nullable?: false;
        auto?: boolean;
    },
): OutputType<string>;
export function relation<T>(
    type: OutputType<T>,
    load: (id: string, ctx: any) => Promise<T | null>,
    {nullable, auto = false}: {
        nullable?: boolean;
        auto?: boolean;
    } = {},
): OutputType<string> {
    return {
        desc: {
            ...type.desc,
            outputExtendable: !auto || undefined,
            nullable,
        },
        resolve: async (id, path, extendQuery, ctx) => {
            if (!auto && !extendQuery) {
                return undefined;
            }
            if (id === null) {
                if (nullable) {
                    return null;
                } else {
                    throw new Error(`Nullable is not allowed but null is passed as id [${path}]`);
                }
            }
            const res = await load(id, ctx);
            if (res === null) {
                if (nullable) {
                    return null;
                } else {
                    throw new Error(`Nullable is not allowed but no record found [${path}]`);
                }
            }
            return type.resolve(
                res,
                path,
                extendQuery,
                ctx,
            );
        },
        checkExtendedQuery: type.checkExtendedQuery || check.optional(check.object({})),
    };
}

export function describe<T>(t: Type<T>, description: string): Type<T>;
export function describe<T>(t: InputType<T>, description: string): InputType<T>;
export function describe<T>(t: OutputType<T>, description: string): OutputType<T>;
export function describe<T>(t: InputType<T> | OutputType<T>, description: string): InputType<T> | OutputType<T> {
    return {
        ...t,
        desc: {
            ...t.desc,
            description,
        },
    };
}

// export function alternatives<Alts extends ReadonlyArray<Type<any>>>(alts: Alts): Type<Alts extends ReadonlyArray<Type<infer T>> ? T : never>;
// export function alternatives<Alts extends ReadonlyArray<InputType<any>>>(alts: Alts): InputType<Alts extends ReadonlyArray<InputType<infer T>> ? T : never>;
// export function alternatives<Alts extends ReadonlyArray<OutputType<any>>>(alts: Alts): OutputType<Alts extends ReadonlyArray<OutputType<infer T>> ? T : never>;
// export function alternatives(alts: any[]) {
//     const res: any = {};
//     if (alts.every(alt => alt.extract)) {
//         const r: InputType<any> = {
//             tsExtract: alts.map(alt => `(${alt.tsExtract})`).join(' | '),
//             extract: check.oneOf(alts.map(alt => alt.extract)),
//         };
//         Object.assign(res, r);
//     }
//     // if (alts.every(alt => alt.resolve)) {
//     //     const r: OutputType<any> = {
//     //         tsResolve: alts.map(alt => `(${alt.tsExtract})`).join(' | '),
//     //         resolve: async (v, path, q, ctx) => Promise.all((v as any[]).map(item => t.resolve(item, `${path}[${item}]`, q, ctx))),
//     //         includeType: t.includeType,
//     //         tsInclude: t.tsInclude,
//     //     };
//     //     Object.assign(res, r);
//     // }
//     return res;
// }