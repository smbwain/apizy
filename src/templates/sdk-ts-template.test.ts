import {arrayOf, boolean, extend, nullable, object, oneOf, optional, relation, string, uuid} from '../type-system';
import {createApi} from '../index';
import {sdkTsTemplate} from './sdk-ts-template';

describe('template', () => {
    describe('ts', () => {
        it('should build', () => {
            interface User {
                id: string;
                email: string;
                nick: string;
            }

            interface Project {
                id: string;
                name: string;
            }

            interface Member {
                id: string;
                name: string;
                email: string | null;
                projectId: string;
                userId: string | null;
                role: 'owner' | 'teacher';
            }

            const api = createApi();

            const $user = api.createEntity<User>('User');
            const $project = api.createEntity<Project>('Project');
            const $member = api.createEntity<Member>('Member');

            $user.addResolver({
                id: uuid(),
                email: string(),
                nick: string(),
            }, (data, ctx) => {
                return {
                    id: data.id,
                    email: data.email,
                    nick: data.nick,
                };
            });
            $project.addResolver({
                id: uuid(),
                name: string(),
            }, (data, ctx) => {
                return {
                    id: data.id,
                    name: data.name,
                };
            });
            $member.addResolver({
                id: uuid(),
                email: nullable(string()),
                name: string(),
                role: oneOf(['owner', 'teacher'] as const),
                project: relation($project, async () => null, {nullable: true}),
                user: relation($user, async () => null, {nullable: true}),
            }, (data) => {
                return {
                    id: data.id,
                    name: data.name,
                    role: data.role,
                    email: data.email,
                    project: data.projectId,
                    user: data.userId,
                };
            });

            api.createMethod('users.getMe', null, nullable($user), async () => {
                return null;
            });
            api.createMethod('projects.get', uuid(), nullable($project), async () => {
                return null;
            });
            api.createMethod(
                'members.list',
                object({
                    projectId: optional(uuid()),
                    userId: optional(uuid()),
                    my: optional(boolean()),
                }),
                arrayOf($member),
                async () => {
                    return [];
                },
            );
            api.createMethod(
                'members.manage.add',
                object({
                    projectId: uuid(),
                    userId: uuid(),
                    name: string(),
                }),
                $member,
                async () => {
                    return null as any;
                },
            );
            api.createMethod(
                'members.manage.update',
                object({
                    id: uuid(),
                    name: optional(string()),
                }),
                $member,
                async () => {
                    return null as any;
                },
            );

            expect(sdkTsTemplate(api.apiDescription)).toMatchSnapshot();
        });
    });
});