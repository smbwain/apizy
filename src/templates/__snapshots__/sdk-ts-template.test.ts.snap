// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template ts should build 1`] = `
"


export type Method<Input, Output, ExtendQuery, FetchOptions> = (
    input: Input,
    options: {
        extend: ExtendQuery;
        fetchOptions: FetchOptions;
    },
) => Promise<Output>;

export type T_User = {
    "$type": "User";
    "id": string;
    "email": string;
    "nick": string;
}
export interface ExtendQuery_User {}

export type T_Project = {
    "$type": "Project";
    "id": string;
    "name": string;
}
export interface ExtendQuery_Project {}

export type T_Member = {
    "$type": "Member";
    "id": string;
    "email": string;
    "name": string;
    "role": "owner" | "teacher";
    "project"?: T_Project;
    "user"?: T_User;
}
export interface ExtendQuery_Member {
    "project"?: ExtendQuery_Project;
    "user"?: ExtendQuery_User;
}

export interface SDK<FetchOptions> {
    "members": {
        "list": Method<{
            "projectId"?: string;
            "userId"?: string;
            "my"?: boolean;
        }, Array<T_Member>, ExtendQuery_Member, FetchOptions>;
        "manage": {
            "add": Method<{
                "projectId": string;
                "userId": string;
                "name": string;
            }, T_Member, ExtendQuery_Member, FetchOptions>;
            "update": Method<{
                "id": string;
                "name"?: string;
            }, T_Member, ExtendQuery_Member, FetchOptions>;
        };
    };
    "projects": {
        "get": Method<string, T_Project, ExtendQuery_Project, FetchOptions>;
    };
    "users": {
        "getMe": Method<null, T_User, ExtendQuery_User, FetchOptions>;
    };
}

export function createSDK<FetchOptions>(
    fetch: (methodName: string, input: any, extend?: any, fetchOptions?: FetchOptions) => Promise<any>,
): SDK<FetchOptions> {
    const createMethod = (methodName: string) => (input: any, options: any) => fetch(methodName, input, options?.extend, options?.fetchOptions);
    return {
        "members": {
            "list": createMethod("members.list"),
            "manage": {
                "add": createMethod("members.manage.add"),
                "update": createMethod("members.manage.update"),
            },
        },
        "projects": {
            "get": createMethod("projects.get"),
        },
        "users": {
            "getMe": createMethod("users.getMe"),
        },
    };
};
"
`;
